File: python/motion_heatmap.py
import os
import collections
import math
import random
import cv2
import scipy.ndimage
import numpy as np


class MotionHeatmap:
    """
    This class is used to generate a motion heatmap from a directory of images representing frames in a time-lapse
    sequence. Instantiate this class with the necessary parameters, then generate the actual motion heatmap using
    the generate_motion_heatmap() method.

    Initialization does the following:
    (1) Divide the input image canvas into the desired number of blocks.
    (2) Pick a random pixel location within each block, to be used as the indicator of motion in that block's region.
    (3) Build a heatmap by quantifying, for each frame, the amount of motion (in the time domain) for each block.
    (4) Construct the average image across all frames in the sequence, if the user specifies this option
    (5) Low-pass filter the heatmap with a Gaussian kernel (with a configurable sigma parameter)
    Generation of the motion heatmap does the following:
    (1) Go through each pixel and determine the heatmap intensity at the block corresponding to the pixel
    (2) Apply positive or negative offsets to the red and blue channels at each pixel depending on the heatmap intensity
    """

    def __init__(
        self,
        # Resolution of the heatmap
        num_vertical_divisions,
        num_horizontal_divisions,
        # Path to images
        image_path,
        # Function or lambda used to filter files in image_path (useful if only certain files in the directory should be chosen)
        file_filter_func=lambda f: True,
        # Set to True to construct an "average" image used for the final overlay image; otherwise, the first frame of the sequence will be used
        use_average_image_overlay=True,
        # Value of sigma for Gaussian filter of the heatmap
        sigma=1.5,
        # Scales intensity of the heatmap's red/blue tint: higher values will exaggerate the tinting
        color_intensity_factor=7,
    ):
        self.num_vertical_divisions = num_vertical_divisions
        self.num_horizontal_divisions = num_horizontal_divisions
        self.color_intensity_factor = color_intensity_factor
        self.use_average_image_overlay = use_average_image_overlay
        self.image_path = image_path

        self.frame_files = filter(
            file_filter_func,
            sorted([f for f in os.listdir(self.image_path) if os.path.isfile(os.path.join(self.image_path, f))]),
        )
        sample_image = cv2.imread(os.path.join(self.image_path, self.frame_files[0]))
        self.height = len(sample_image)
        self.width = len(sample_image[0])

        # Initialize random pixel locations for each block
        self.pixel_locations = {}
        for row in range(self.num_vertical_divisions):
            for col in range(self.num_horizontal_divisions):
                self.pixel_locations[(row, col)] = (
                    int(row * self.height/num_vertical_divisions + math.floor(random.random() * self.height/num_vertical_divisions)),
                    int(col * self.width/num_horizontal_divisions + math.floor(random.random() * self.width/num_horizontal_divisions)),
                )

        # Initialize block intensities for each block and construct average image
        self.block_intensities = collections.defaultdict(list)
        self.average_image = np.zeros((self.height, self.width, 3))
        for index, file_name in enumerate(self.frame_files):
            print 'Processing input frame {index} of {total}'.format(index=index + 1, total=len(self.frame_files))
            frame = cv2.imread(os.path.join(self.image_path, file_name), cv2.IMREAD_COLOR)
            if self.use_average_image_overlay:
                self.average_image += frame
            for row in range(self.num_vertical_divisions):
                for col in range(self.num_horizontal_divisions):
                    pixel_row, pixel_col = self.pixel_locations[(row, col)]
                    self.block_intensities[(row, col)].append(round(np.mean(frame[pixel_row][pixel_col])))
        if self.use_average_image_overlay:
            self.average_image /= len(self.frame_files)

        # Calculate heatmap
        unfiltered_heatmap = np.zeros((self.num_vertical_divisions, self.num_horizontal_divisions))
        for row in range(self.num_vertical_divisions):
            for col in range(self.num_horizontal_divisions):
                unfiltered_heatmap[row][col] = np.std(self.block_intensities[(row, col)])
        self.heatmap = scipy.ndimage.filters.gaussian_filter(unfiltered_heatmap, sigma=sigma)

    def generate_motion_heatmap(self, file_name='motion_heatmap.jpg'):
        """
        Generates the motion heatmap for this sequence of frames.

        :param file_name: The name of the file to save to disk.
        :return: True or False denoting success of the operation. The method will write the output image to disk.
        """
        output_image = self.average_image if self.use_average_image_overlay else cv2.imread(os.path.join(self.image_path, self.frame_files[0]), cv2.IMREAD_COLOR)
        mean_stdev = np.mean(self.heatmap)

        for vertical_index in range(self.num_vertical_divisions):
            for horizontal_index in range(self.num_horizontal_divisions):
                print 'Processing output block {index} of {total}'.format(
                    index=vertical_index * self.num_horizontal_divisions + horizontal_index + 1,
                    total=self.num_horizontal_divisions * self.num_vertical_divisions,
                )
                offset = self.color_intensity_factor * (self.heatmap[vertical_index][horizontal_index] - mean_stdev)
                for i in range(self.height/self.num_vertical_divisions):
                    for j in range(self.width/self.num_horizontal_divisions):
                        row = vertical_index * self.height/self.num_vertical_divisions + i
                        col = horizontal_index * self.width/self.num_horizontal_divisions + j
                        # Apparently indices [0 1 2] correspond to [B G R].
                        # Why, OpenCV?
                        output_image[row][col][2] = self._clip_rgb(output_image[row][col][2] + offset)
                        output_image[row][col][0] = self._clip_rgb(output_image[row][col][0] - offset)

        return cv2.imwrite(file_name, output_image)

    @staticmethod
    def _clip_rgb(value):
        """
        Clips the value to be within the bounds of an 8-bit color image, e.g. in the range [0, 255].

        :param value: The value to clip.
        :return: An integer of the value clipped, as necessary.
        """
        return int(max(min(value, 255), 0))

File: matlab/motion_heatmap.m
% CONFIGURATION
% Resolution of the heatmap
num_vertical_divisions = 36;
num_horizontal_divisions = 64;
% Scales intensity of the heatmap's red/blue tint: higher values will
% exaggerate the tinting
color_intensity_factor = 7;
% Set to true to construct an "average" image used for the final overlay
% image. Otherwise, the first frame of the sequence will be used
use_average_image_overlay = 1;

fprintf('Loading images and initializing\n');
frame_files = dir('*.jpg');
[height, width, null] = size(imread(frame_files(1).name));  % Dimensions of images
[num_frames, null] = size(frame_files);  % Number of total frames
num_blocks = num_horizontal_divisions * num_vertical_divisions;
image_locations = zeros(1, 2, num_blocks);  % Stores pixel locations of each block
intensities = zeros(num_blocks, num_frames);  % Vertical index corresponds to block; horizonal index corresponds to intensity for a pixel within that block
average_image = zeros(height, width, 3);
average_image_red = zeros(height, width);
average_image_green = zeros(height, width);
average_image_blue = zeros(height, width);

% Select a random pixel location for each block and keep track of it
% This allows us to imread each image once, and examine the same pixel
% location every time for each new image
fprintf('Selecting random pixel locations for each block; total %i\n', num_blocks);
for block_index = 1:num_blocks
    block_row = ceil(block_index/num_horizontal_divisions);
    block_col = block_index - num_horizontal_divisions * (block_row - 1);
    random_row = (block_row - 1) * (height/num_vertical_divisions) + floor(rand * height/num_vertical_divisions) + 1;
    random_col = (block_col - 1) * (width/num_horizontal_divisions) + floor(rand * width/num_horizontal_divisions) + 1;
    image_locations(1, :, block_index) = [random_row, random_col];
end

% Iterate through every frame and keep track of the grayscale intensity at
% the randomly selected pixel above. Also construct an average image, if
% enabled.
for frame_index = 1:num_frames
    fprintf('Processing frame %i of %i\n', frame_index, num_frames);
    image = imread(frame_files(frame_index).name);
    if (use_average_image_overlay)
        average_image_red = average_image_red + double(image(:, :, 1))/num_frames;
        average_image_green = average_image_green + double(image(:, :, 2))/num_frames;
        average_image_blue = average_image_blue + double(image(:, :, 3))/num_frames;
    end
    image = double(rgb2gray(image));
    for block_index = 1:num_blocks
        pixel = image_locations(1, :, block_index);
        intensities(block_index, frame_index) = image(pixel(1), pixel(2));
    end
end
if (use_average_image_overlay)
    average_image(:, :, 1) = average_image_red;
    average_image(:, :, 2) = average_image_green;
    average_image(:, :, 3) = average_image_blue;
end

% Generate a heatmap by considering the standard deviation of the signal at
% each block
fprintf('Generating standard deviation heatmap\n');
heatmap = zeros(num_vertical_divisions, num_horizontal_divisions);
for block_index = 1:num_blocks
    block_row = ceil(block_index/num_horizontal_divisions);
    block_col = block_index - num_horizontal_divisions * (block_row - 1);
    heatmap(block_row, block_col) = std(intensities(block_index, :));
end
heatmap_filt = imgaussfilt(heatmap, 1.5);
figure, mesh(heatmap_filt);
title(sprintf('Time-averaged spatial concentration of motion; Gaussian-filtered (\\sigma = 1.5, N = %i)', num_frames));
xlabel('Horizontal block index');
ylabel('Vertical block index');
zlabel('Standard deviation of grayscale intensity');

% Create the final output image by adjusting the RGB channel values of each
% pixel according to the heatmap's value
fprintf('Creating output image with heatmap overlay\n');
mean_std = mean(mean(heatmap_filt));
if (use_average_image_overlay)
    overlay_image = uint8(average_image);
else
    overlay_image = imread(frame_files(1).name);
end
output_image = overlay_image;
for i = 1:height
    for j = 1:width
        vertical_index = ceil(num_vertical_divisions * i/height);
        horizontal_index = ceil(num_horizontal_divisions * j/width);
        output_image(i, j, :) = overlay_image(i, j, :);
        output_image(i, j, 1) = overlay_image(i, j, 1) + color_intensity_factor*(heatmap_filt(vertical_index, horizontal_index) - mean_std);
        if (output_image(i, j, 1) > 255)
            output_image(i, j, 1) = 255;
        end
        output_image(i, j, 3) = overlay_image(i, j, 3) - color_intensity_factor*(heatmap_filt(vertical_index, horizontal_index) - mean_std);
        if (output_image(i, j, 3) < 0)
            output_image(i, j, 3) = 0;
        end
    end
end
figure, imshow(output_image);
title('Motion heatmap');